File: principal-engineer/java/src/main/java/com/interview/order/service/ServiceResult.java
package com.interview.order.service;

public class ServiceResult<T> {
    private final int status;
    private final T value;
    private final String error;

    private ServiceResult(int status, T value, String error) {
        this.status = status;
        this.value = value;
        this.error = error;
    }

    public static <T> ServiceResult<T> ok(T value) {
        return new ServiceResult<>(200, value, null);
    }

    public static <T> ServiceResult<T> created(T value) {
        return new ServiceResult<>(201, value, null);
    }

    public static <T> ServiceResult<T> badRequest(String error) {
        return new ServiceResult<>(400, null, error);
    }

    public static <T> ServiceResult<T> notFound(String error) {
        return new ServiceResult<>(404, null, error);
    }

    public static <T> ServiceResult<T> conflict(String error) {
        return new ServiceResult<>(409, null, error);
    }

    public int getStatus() {
        return status;
    }

    public T getValue() {
        return value;
    }

    public String getError() {
        return error;
    }
}

File: principal-engineer/csharp/src/OrderService/Services/ServiceResult.cs
namespace OrderService.Services;

/// <summary>
/// Transport-agnostic service result. The HTTP layer maps <see cref="Status"/>
/// to a status code.
/// </summary>
public record ServiceResult<T>(int Status, T? Value = default, string? Error = null)
{
    public static ServiceResult<T> Ok(T value) => new(200, value);
    public static ServiceResult<T> Created(T value) => new(201, value);
    public static ServiceResult<T> NoContent() => new(204);
    public static ServiceResult<T> BadRequest(string message) => new(400, default, message);
    public static ServiceResult<T> NotFound(string message = "Not found") => new(404, default, message);
    public static ServiceResult<T> Conflict(string message) => new(409, default, message);
}
